lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1577,10 +1642,22 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       // Until Swift 6, only emit a warning when we get this with an
       // explicit capture, since we used to not diagnose this at all.
       auto shouldOnlyWarn = [&](Expr *selfRef) {
-        // We know that isImplicitSelfParamUseLikelyToCauseCycle is true,
-        // which means all these casts are valid.
-        return !cast<VarDecl>(cast<DeclRefExpr>(selfRef)->getDecl())
-                  ->isSelfParameter();
+        if (auto declRef = dyn_cast<DeclRefExpr>(selfRef)) {
+          if (auto decl = declRef->getDecl()) {
+            if (auto varDecl = dyn_cast<VarDecl>(decl)) {
+              // If the self decl was defined in an `if` or `guard` statement, we know this is an inner closure of some outer closure's `weak self` capture. Since this wasn't allowed in Swift 5.5, we should just always emit an error.
+              if (auto parentStmt = varDecl->getParentPatternStmt()) {
+                if (isa<GuardStmt>(parentStmt) || isa<IfStmt>(parentStmt)) {
```

Comment:
Something to consider (this should be discussed on evolution): Why not allow all labelled control flow statements? This allows `while let self = self { /**/ }` to go through.

Reply:
This makes sense -- no reason to not include `while` here.